### PR TITLE
Feature/timestamp-interval arith query

### PR DESCRIPTION
**Pull request by berkaysynnada**
# Which issue does this PR close?

Closes #5704 and #194.

# Rationale for this change

We can handle such queries now:
`SELECT val, ts1 - ts2 AS ts_diff FROM table_a`
`SELECT val, interval1 - interval2 AS interval_diff FROM table_a`
`SELECT val, ts1 - interval1 AS ts_interval_diff FROM table_a`
`SELECT val, interval1 + ts1 AS interval_ts_sub FROM table_a`

# What changes are included in this PR?

  | - | +
-- | -- | --
timestamp op timestamp (same type) | OK: second and millisecond types give results in daytime(day+millisecond), microsecond and nanosecond types give result in monthdaynano(month+day+nano, but month field is not used) | N/A
timestamp op timestamp (different types) | N/A | N/A
interval op interval (same type) | OK: operations are done field by field, gives the same type | OK: operations are done field by field, gives the same type
interval op interval (different types) | OK: give result in monthdaynano type | OK: give result in monthdaynano type
timestamp op interval | OK: give result in the type of the timestamp | OK: give result in the type of the timestamp
interval op timestamp | N/A | OK: the same of timestamp + interval 

Some match expressions in planner.rs, binary.rs, and datetime.rs are extended. Coerced types and allowable operations are shown in the table. 

I try to use existing scalar value functions as much as possible to not duplicate. However, in arrow.rs, subtraction and addition functions are for numeric types, hence I need to add some functions to call with `binary` function. 

In datetime.rs, the `evaluate` function was written to accept only "Array + Scalar" or "Scalar + Scalar" values to evaluate. It is extended to accept "Array + Array", and 4 different variations of that case (Timestamp op Timestamp, Interval op Interval, Timestamp op Interval, Interval op Timestamp) are implemented. "Array + Scalar" evaluations are done with `unary` function in arrow.rs, and I follow the similar pattern with `binary` function.

# Are these changes tested?

Yes, there are tests for each match in timestamp.slt. However, tables with intervals cannot be created like `INTERVAL '1 second'`, since some work is needed in arrow-rs for casting. Timestamp difference case with timezone is also left in timestamp.rs because of a similar reason.

# Are there any user-facing changes?

### CODE REVIEW

File: datafusion/core/tests/sql/mod.rs [hunk 1]
```diff
@@ -1359,6 +1359,176 @@ where
...
+
+    let divisor = match A::get_time_unit() {
+        TimeUnit::Nanosecond => 1,
+        TimeUnit::Microsecond => 1000,
+        TimeUnit::Millisecond => 1_000_000,
+        TimeUnit::Second => 1_000_000_000,
+    };
+
+    let timestamps1 = vec![
+        1_678_892_420_000_000_000i64 / divisor,
```
Comment (mustafasrepo) [inline datafusion/core/tests/sql/mod.rs:1398]: Add a comment for what these numbers represent

File: datafusion/core/tests/sql/mod.rs [hunk 2]
```diff
@@ -1359,6 +1359,176 @@ where
...
+        ),
+        Field::new(
+            "interval2",
+            DataType::Interval(IntervalUnit::DayTime),
+            false,
+        ),
+        Field::new("val", DataType::Int32, true),
+    ]));
+
+    let intervals1 = vec![4_394_969_299, 4_494_969_298, 4_594_969_297];
```
Comment (mustafasrepo) [inline datafusion/core/tests/sql/mod.rs:1440]: Add a comment for what these numbers represent

File: datafusion/core/tests/sql/mod.rs [hunk 3]
```diff
@@ -1359,6 +1359,176 @@ where
     Ok(Arc::new(table))
 }
 
+fn make_timestamp_sub_table<A>() -> Result<Arc<MemTable>>
```
Comment (mustafasrepo) [inline datafusion/core/tests/sql/mod.rs:1362]: I think `make_interval_ts_add_table` and `make_ts_interval_sub_table` are same. Delete one of them and change the name of remaining one.

File: datafusion/physical-expr/src/expressions/datetime.rs [hunk 4]
```diff
@@ -59,27 +56,29 @@ impl DateTimeIntervalExpr {
...
-                    ))),
-                }
-            }
+        match (
+            lhs.data_type(input_schema)?,
+            op,
+            rhs.data_type(input_schema)?,
+        ) {
+            (
+                DataType::Date32 | DataType::Date64 | DataType::Timestamp(_, _),
```
Comment (mustafasrepo) [inline datafusion/physical-expr/src/expressions/datetime.rs:65]: You can use is_date, is_timestamp() etc in here.

File: datafusion/physical-expr/src/expressions/datetime.rs [hunk 5]
```diff
@@ -142,14 +135,30 @@ impl PhysicalExpr for DateTimeIntervalExpr {
                 return Err(DataFusionError::Internal(msg.to_string()));
             }
         };
+        // RHS is first checked. If it is a Scalar, there are 2 options:
+        // Either LHS is also a Scalar and matching operation is applied,
+        // or LHS is an Array and unary operations for related types are
+        // applied in evaluate_array function. If RHS is an Array, then
+        // LHS must also be, moreover; they must be the same Timestamp type.
+        match (rhs_value, lhs_value) {
```
Comment (mustafasrepo) [inline datafusion/physical-expr/src/expressions/datetime.rs:143]: Is is better to match `(lhs_value, rhs_value)` I think. Is there any reason to do this?

File: datafusion/physical-expr/src/expressions/datetime.rs [hunk 6]
```diff
@@ -239,6 +248,486 @@ pub fn evaluate_array(
...
+            array_rhs.data_type()
+        ))),
+    }
+}
+
+fn interval_array_op(
+    array_lhs: &ArrayRef,
+    array_rhs: &ArrayRef,
+    sign: i32,
+) -> Result<ArrayRef, DataFusionError> {
```
Comment (mustafasrepo) [inline datafusion/physical-expr/src/expressions/datetime.rs:432]: make this `Result<ArrayRef>`

File: datafusion/physical-expr/src/expressions/datetime.rs [hunk 7]
```diff
@@ -239,6 +248,486 @@ pub fn evaluate_array(
...
+            array_rhs.data_type()
+        )))?,
+    };
+    Ok(ColumnarValue::Array(ret))
+}
+
+fn ts_array_op(
+    array_lhs: &ArrayRef,
+    array_rhs: &ArrayRef,
+) -> Result<ArrayRef, DataFusionError> {
```
Comment (mustafasrepo) [inline datafusion/physical-expr/src/expressions/datetime.rs:358]: make this `Result<ArrayRef>`

File: datafusion/physical-expr/src/expressions/datetime.rs [hunk 8]
```diff
@@ -239,6 +249,496 @@ pub fn evaluate_array(
...
+    }};
+}
+macro_rules! ts_interval_op {
+    ($lhs:ident, $rhs:ident, $caster1:expr, $caster2:expr, $op:expr, $sign:ident, $type_in1:ty, $type_in2:ty, $mode:expr) => {{
+        let prim_array_lhs = $caster1(&$lhs)?;
+        let prim_array_rhs = $caster2(&$rhs)?;
+        let ret = Arc::new(binary::<$type_in1, $type_in2, _, $type_in1>(
+            prim_array_lhs,
+            prim_array_rhs,
+            |ts, interval| $op(ts, interval as i128, $sign, $mode).unwrap(),
```
Comment (metesynnada) [inline datafusion/physical-expr/src/expressions/datetime.rs:312]: Can we handle `.unwrap()` as returning `Result`?

File: datafusion/physical-expr/src/expressions/datetime.rs [hunk 9]
```diff
@@ -239,6 +249,496 @@ pub fn evaluate_array(
...
+                            &$lhs_tz,
+                        )
+                        .expect("{ts1} timestamp cannot build a DateTime object"),
+                    ),
+                    $ts_unit(
+                        &with_timezone_to_naive_datetime::<$mode>(
+                            ts2.mul_wrapping($coef),
+                            &$rhs_tz,
+                        )
+                        .expect("{ts2} timestamp cannot build a DateTime object"),
```
Comment (metesynnada) [inline datafusion/physical-expr/src/expressions/datetime.rs:273]: Can we return error instead of handling it here?

**Comment by berkaysynnada** [comment]
> FYI I think `timestamp - timestamp` should result in an `interval` (not a `timestamp`)
> 
> Also, `timestamp + timestamp` I don't think makes logical sense and thus should error

"timestamp op timestamp (same type)" means the timestamps having the same type (like millisecond and millisecond) are  the inputs of the operation. They already result in an interval.

"timestamp + timestamp" operation gives an error in the function `temporal_add_sub_coercion`, saying it is an unsupported operation.

**Pull request closed without merging**